url9e7e48285f_nalbachk_MidiToPhaseShift_tgz-pJ8-com_my_main_MainJ8/src/com/my/converter/TrackConverterStringsReal.java

```diff
@@ -0,0 +1,166 @@
+
+										tabPosition = tabTrack.getPositionByIndex(tabPositionIndex);
+										notePhaseShiftFromTab = getNotePhaseShiftFromTab(tabPosition, notePhaseShift);
+									} while (null == notePhaseShiftFromTab);
+								}
+
+								// take note from tab
+								if (null != notePhaseShiftFromTab) {
+									notePhaseShift = notePhaseShiftFromTab;
+								}
+							}
+
+							NoteEof noteEof = config.getNoteEof(notePhaseShift);
+							Integer data1 = getData1(noteEof);
+							Integer data2 = getData2(noteEof);
+							shortMessage.setMessage(shortMessage.getStatus(), data1, data2);
+							break;
+					}
+				}
+			}
+		} catch (InvalidMidiDataException e) {
+			LOG.error("cannot modify trackNotes");
+		}
+	}
+
+	public int getData1(NoteEof noteEof) {
+		return 95 + noteEof.getLineNr();
+	}
+
+	public int getData2(NoteEof noteEof) {
+		return 100 + noteEof.getFret();
+	}
+
+	protected NotePhaseShift getNotePhaseShiftFromTab(TabPosition tabPosition, NotePhaseShift notePhaseShift) {
+		Integer midiNoteFromMapping = config.getNoteMidi(notePhaseShift);
+		for (NotePhaseShift fromTab : tabPosition.getNotes()) {
+			if (null != fromTab) {
+				Integer midiNoteFromTab = config.getNoteMidi(fromTab);
+				if (midiNoteFromTab == midiNoteFromMapping) { // midi note value is the same
+					return fromTab;
+				}
+			}
+		}
+
+		// no tab note found
+		return null;
+	}
+
+	private void removeMidiEvents(Track midiTrack) {
+		List<MidiEvent> midiEventsToRemove = new ArrayList<MidiEvent>();
+
+		midiEventsToRemove.addAll(this.getMidiEventsToRemoveByStatus(midiTrack));
+
+		LOG.info("midiEventsToRemove: {}", midiEventsToRemove.size());
+		for (MidiEvent midiEvent : midiEventsToRemove) {
+			midiTrack.remove(midiEvent);
+		}
+	}
+
+	private List<MidiEvent> getMidiEventsToRemoveByStatus(Track midiTrack) {
+		Map<Integer, List<MidiEvent>> midiEventsPerStatus = new HashMap<>();
+
+		for (int index = 0; index < midiTrack.size(); index++) {
+			MidiEvent midiEvent = midiTrack.get(index);
+			MidiMessage midiMessage = midiEvent.getMessage();
+
+			if (midiMessage instanceof ShortMessage) {
+				ShortMessage shortMessage = (ShortMessage) midiMessage;
+
+				switch (shortMessage.getCommand()) {
+					case ShortMessage.NOTE_ON:
+						// break;
+					case ShortMessage.NOTE_OFF:
+						Integer status = shortMessage.getStatus();
+						if (!midiEventsPerStatus.containsKey(status)) {
+							midiEventsPerStatus.put(status, new ArrayList<MidiEvent>());
+						}
+						midiEventsPerStatus.get(status).add(midiEvent);
+						break;
+				}
+			}
+		}
+
+		// if less then X midi events with same state => remove them
+		List<MidiEvent> midiEventsToRemove = new ArrayList<MidiEvent>();
+		for (Integer status : midiEventsPerStatus.keySet()) {
+			List<MidiEvent> midiEvents = midiEventsPerStatus.get(status);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.